### PR TITLE
Initial skeleton

### PR DESCRIPTION
**Pull request by kemurphy**
General shape for decoupling WS events from internal events and for preventing simultaneous logins

### CODE REVIEW

File: src/hanabi/conn.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    let ws = tokio_tungstenite::accept_hdr_async(stream, accept_cb)
+        .await
+        .context("Handshake error")?;
+
+    info!("New WebSocket connection: {}", addr);
+
+    let session = HanabiSession::new(user_id.unwrap());
+
+    Ok(session::run(session, ws).await?)
```
Comment (sjdrodge) [inline src/hanabi/conn.rs:92]: Is there a difference between this and just a plain `session::run(session, ws).await?`

File: src/session/mod.rs [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                                    handoff_complete,
+                                });
+
+                            let (packet, err) = match packet {
+                                Ok(packet) => (Some(packet), Ok(())),
+                                Err(_) => (None, Err(LifecycleError::MapUpdaterClosed)),
+                            };
+
+                            err.and(chan.send(packet).map_err(|_| LifecycleError::HandoffChannelClosed))?;
+                            return Ok(());
```
Comment (sjdrodge) [inline src/session/mod.rs:119]: shouldn't this also be `break Ok(())`?

File: src/session/mod.rs [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+                        Event::Message(msg) => {
+                            self.session.handle_message(&mut self.sink, msg).await;
+                        },
+                        Event::Event(e) => {
+                            let _ = e;
+                            unimplemented!()
+                        },
+                        Event::Handoff(chan) => {
+                            // First, make anyone that asks for this event sink wait for handoff.
+                            let handoff_complete = self.notify.clone();
```
Comment (sjdrodge) [inline src/session/mod.rs]: I prefer using `Arc::clone()` instead of `.clone()` here.

File: src/hanabi/conn.rs (hunk 1 above)
Comment (kemurphy) [reply src/hanabi/conn.rs:92]: `?` doesn't Ok-wrap, so it'll be a type error (`expected anyhow::Result<()>, got ()` or somesuch).

File: src/session/mod.rs (hunk 2 above)
Comment (kemurphy) [reply src/session/mod.rs:119]: I think the others should be `return` instead -- I'll make them consistent.

**Pull request merged**